internal/configurator/controller.go

```diff
@@ -0,0 +1,178 @@
+	})
+
+	// Watch all the endpoints.
+	endpointsInformer := informerFactory.Core().V1().Endpoints()
+	endpointsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(_ interface{}) { q.AddRateLimited(struct{}{}) },
+		// We rate limit the update func on endpoints, which can generate a lot of noise, when deploying pods.
+		UpdateFunc: func(_, obj interface{}) { q.AddRateLimited(struct{}{}) },
+		DeleteFunc: func(_ interface{}) { q.AddRateLimited(struct{}{}) },
+	})
+
+	// Watch the mesh nodes pods.
+	podsInformer := informerFactory.Core().V1().Pods()
+	podsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { enqueuePodEvent(q, obj, l) },
+		UpdateFunc: func(_, obj interface{}) { enqueuePodEvent(q, obj, l) },
+		DeleteFunc: func(obj interface{}) { enqueuePodEvent(q, obj, l) },
+	})
+
+	return &Controller{
+		deployer:         deployer,
+		provider:         provider,
+		pods:             podsInformer.Lister(),
+		queue:            q,
+		exited:           make(chan struct{}),
+		currentNamespace: currentNamespace,
+		logger:           l.WithField("module", moduleName),
+	}
+}
+
+// Run process all events coming from the workqueue until shutdown is signaled.
+func (c *Controller) Run() {
+	defer close(c.exited)
+
+	for c.processEvent() {
+	}
+}
+
+// ShutDown shutdowns the controller.
+func (c *Controller) ShutDown() {
+	c.queue.ShutDown()
+}
+
+// Wait waits for the termination of the controller loop.
+func (c *Controller) Wait(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-c.exited:
+		return nil
+	}
+}
+
+func (c *Controller) processEvent() bool {
+	item, shutdown := c.queue.Get()
+	if shutdown {
+		return false
+	}
+
+	defer c.queue.Done(item)
+
+	cfg, err := c.provider.BuildConfig()
+	if err != nil {
+		c.handleError(item, err)
+		return true
+	}
+
+	maeshNodes, err := c.pods.Pods(c.currentNamespace).List(resource.MeshPodsLabelsSelector())
+	if err != nil {
+		c.handleError(item, err)
+		return true
+	}
+
+	if err = c.deployer.Deploy(maeshNodes, cfg); err != nil {
+		c.handleError(item, err)
+		return true
+	}
+
+	c.queue.Forget(item)
+
+	return true
+}
+
+const (
+	maxAttempts = 3
+)
+
+func (c *Controller) handleError(item interface{}, err error) {
+	l := c.logger.WithError(err)
+
+	attempt := c.queue.NumRequeues(item)
+	if attempt < maxAttempts {
+		l.Errorf("(%d/%d) unable to process item, retrying", attempt+1, maxAttempts)
+		c.queue.AddRateLimited(item)
+		return
+	}
+
+	l.Errorf("Max attempt reached, ignoring failing event")
+	c.queue.Forget(item)
+}
+
+func enqueueSvcEvent(q workqueue.RateLimitingInterface, obj interface{}, log logrus.FieldLogger) {
+	isMeshSvc, err := resource.IsMeshService(obj)
+	if err != nil {
+		log.WithError(err).Error("unable to determine if event comes from a mesh service, skipping")
+		return
+	}
+	if !isMeshSvc {
+		return
+	}
+
+	q.AddRateLimited(struct{}{})
```

Comment:
This is mysterious to me 🤔 Why do we add an empty struct to the rate limiting work queue? What does that do? Could we add a comment here to explain it?

Reply:
This controller doesn't really care about what really happened, it just needs to recalculate dynamic configuration from current cache. So we don't provide any information about this event, only that it happened. Thus the empty struct.